dht_rt_seeder.go

```diff
@@ -0,0 +1,111 @@
+package dht
+
+import (
+	"context"
+
+	"github.com/ipfs/go-todocounter"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+func (dht *IpfsDHT) seedRoutingTable(candidates, fallbacks []peer.ID) error {
+	peersFromSet := func(m map[peer.ID]struct{}) []peer.ID {
+		peers := make([]peer.ID, len(m))
+		i := 0
+		for p := range m {
+			peers[i] = p
+			i++
+		}
+
+		return peers
+	}
+
+	// candidates we have still not tried dialling to
+	remainingCandidates := make(map[peer.ID]struct{})
+	// fallbacks we have still not tried dialling to
+	remainingFallbacks := make(map[peer.ID]struct{})
+
+	for i := range candidates {
+		remainingCandidates[candidates[i]] = struct{}{}
+	}
+
+	for i := range fallbacks {
+		remainingFallbacks[fallbacks[i]] = struct{}{}
+	}
+
+	for {
+
+		todial := dht.seedsProposer.Propose(dht.routingTable, peersFromSet(remainingCandidates), peersFromSet(remainingFallbacks))
+		if len(todial) == 0 {
+			logger.Infof("dht rt seeder: finished as proposer returned nil, RT size is now %d", dht.routingTable.Size())
+			return nil
+		} else {
+			// start dialing
+			type result struct {
+				p   peer.ID
+				err error
+			}
+
+			// attempts to dial to a given peer to verify it's available
+			dialFn := func(ctx context.Context, p peer.ID, res chan<- result) {
+				childCtx, cancel := context.WithTimeout(ctx, dht.seederDialTimeout)
+				defer cancel()
+				_, err := dht.host.Network().DialPeer(childCtx, p)
+				select {
+				case <-ctx.Done(): // caller has already hung up & gone away
+				case res <- result{p, err}:
+				}
+			}
+
+			resCh := make(chan result) // dial results.
+			ctx, cancel := context.WithTimeout(dht.ctx, dht.totalSeederDialTimeout)
+			defer cancel()
+
+			// start dialing
+			semaphore := make(chan struct{}, dht.seederConcurrentDials)
+			go func(peers []peer.ID) {
+				for _, p := range peers {
+					semaphore <- struct{}{}
+					go func(p peer.ID, res chan<- result) {
+						dialFn(ctx, p, res)
+						<-semaphore
+					}(p, resCh)
+				}
+			}(todial)
+
+			// number of results we expect on the result channel
+			todo := todocounter.NewSyncCounter()
+			todo.Increment(uint32(len(todial)))
+
+		LOOP:
+			for {
+				select {
+				case res := <-resCh:
```

Comment:
Why do we wait for all the dials to be done before asking for more proposed peers to dial, doesn't this throttle how many peers we can dial simultaneously?

Reply:
This problem has been fixed now.